### PR TITLE
[WIP] Fixes #181 and adds support for OAuth2 / Okta. 

### LINKED ISSUE #181: Support existing Keycloak instance
Thanks for the OIDC plugin and including a setup for Keycloak in this helm chart! 

For those of us who already have existing Keycloak instances, it would be great to be able to leverage those by configuring the Keycloak component to point to our existing instance rather than deploying a new one.

### PR DESCRIPTION
**Pull request by devinbost**
This PR Fixes #181 and provides additional examples for using TLS with S4K and OAuth2.

Also, it's a WIP until I finish doing some testing. 

Remaining items:
1. Resolve auth failures from Prometheus. This requires adding an [OAuth2 section](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.OAuth2) to the PodMonitors provided [in this PR](https://github.com/datastax/pulsar-helm-chart/pull/270) once it's merged into master. 
2. (Optional) Resolve broker restart that occurs when `attemptAuthenticationProviderToken: "true"`. See note: https://github.com/datastax/pulsar-openid-connect-plugin#using-authenticationprovidertoken
3. Investigate/resolve auth issues with S4K

### CODE REVIEW

File: helm-chart-sources/pulsar/Chart.yaml [hunk 1]
```diff
@@ -23,10 +23,11 @@ maintainers:
   - name: cdbartholomew
   - name: lhotari
   - name: michaeljmarshall
-version: 3.0.4
+  - name: devinbost
+version: 3.1.0
 dependencies:
 - name: kube-prometheus-stack
-  version: 35.x.x
+  version: 41.x.x
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/Chart.yaml:30]: This should be upgraded in its own PR.

File: helm-chart-sources/pulsar/Chart.yaml [hunk 2]
```diff
@@ -23,10 +23,11 @@ maintainers:
   - name: cdbartholomew
   - name: lhotari
   - name: michaeljmarshall
-version: 3.0.4
+  - name: devinbost
+version: 3.1.0
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/Chart.yaml:27]: We bump the version at release time.

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml [hunk 3]
```diff
@@ -81,17 +81,52 @@ data:
...
+  openIDAcceptedTimeLeewaySeconds: "0"
+  openIDJwkCacheSize: "10"
+  openIDJwkExpiresSeconds: "300"
+  openIDJwkConnectionTimeoutMillis: "10000"
+  openIDJwkReadTimeoutMillis: "10000"
+  openIDMetadataCacheSize: "10"
+  openIDMetadataExpiresSeconds: "86400"
+  openIDMetadataConnectionTimeoutMillis: "10000"
+  openIDMetadataReadTimeoutMillis: "10000"
+  openIDRequireHttps: "true"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml:124]: We could probably set these once in the values file and then let users override them. I'd expect them to be the same for every component, so no need to make the config per component.

Also, all of these need to be prefixed with `PULSAR_PREFIX_`.

File: helm-chart-sources/pulsar/values.yaml [hunk 4]
```diff
@@ -1750,6 +1750,21 @@ pulsarHeartbeat:
...
+  enabled: false # If true, make sure tokenAuth is disabled.
+  allowedAudience: api://pulsarClient
+  roleClaim: sub
+  attemptAuthenticationProviderToken: true
+  authenticationProvider: com.datastax.oss.pulsar.auth.AuthenticationProviderOpenID
+  # From token generated in Okta UI or other method:
+  issuerUrl: https://example.okta.com/oauth2/au...97 
+  scope: pulsar_client_m2m
+  client_id: "0oa..."
+  client_secret: "04g3P..."
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1766]: Nit: I think these defaults should just be the empty string.

File: helm-chart-sources/pulsar/values.yaml [hunk 5]
```diff
@@ -1750,6 +1750,21 @@ pulsarHeartbeat:
...
+# Deploy OpenID (set to false if Keycloak is enabled)
+# For configuring OpenID in Okta for Pulsar, instructions are here: https://www.youtube.com/watch?v=UQBrecHOXxU&ab_channel=DataStaxDevelopers
+# Then, use the values from the Payload generated to populate these fields:
+openid:
+  enabled: false # If true, make sure tokenAuth is disabled.
+  allowedAudience: api://pulsarClient
+  roleClaim: sub
+  attemptAuthenticationProviderToken: true
+  authenticationProvider: com.datastax.oss.pulsar.auth.AuthenticationProviderOpenID
+  # From token generated in Okta UI or other method:
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1762]: Maybe we should have a section in the readme describing the relevant okta settings? Also, this is not from the token, this is where users retrieve the token.

File: helm-chart-sources/pulsar/templates/utils/health-configmap.yaml [hunk 6]
```diff
@@ -31,6 +31,12 @@ data:
     #!/bin/bash
     {{- if .Values.enableTokenAuth }}
     curl -s --max-time {{ .Values.proxy.probe.timeout }} --fail -H "Authorization: Bearer $(cat /pulsar/token-superuser/superuser.jwt | tr -d '\r')" http://localhost:8080/metrics/ > /dev/null
+    {{- else if .Values.openid.enabled }}
+    # Passing credentials in format client_id:client_secret using basic auth:
+    TOKEN="{{ .Values.openid.client_id }}":{{ .Values.openid.client_secret }}" | base64
+    # NOTE: It's probably insecure to store these credentials in here. What's the better way to do this?
+    #       We could copy the client_id and client_secret to separate Secret object entries and pull them that way, perhaps.
+    curl -s --max-time {{ .Values.proxy.probe.timeout }} --fail -H "Authorization: Basic $TOKEN" http://localhost:8080/metrics/ > /dev/null
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/utils/health-configmap.yaml:39]: As we discussed offline, this approach won't work. I think we'll want to integrate long lived token auth and oauth2 for now.

**Review by michaeljmarshall — COMMENTED** [review]
Great progress @devinbost! Let me know what questions you have.

File: helm-chart-sources/pulsar/templates/utils/health-configmap.yaml (hunk 6 above)
Comment (devinbost) [reply helm-chart-sources/pulsar/templates/utils/health-configmap.yaml:39]: haha sorry, I thought I removed this change. I'm glad you caught it.

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml (hunk 3 above)
Comment (devinbost) [reply helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml]: @michaeljmarshall PTAL.

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    PULSAR_MEM: "-Xms200m -Xmx200m -XX:MaxDirectMemorySize=56m"
+    PULSAR_EXTRA_OPTS: "-Dpulsar.log.root.level=info"
+    PULSAR_PREFIX_kafkaListeners: "SASL_PLAINTEXT://0.0.0.0:9092,SASL_SSL://0.0.0.0:9093"
+    # here you have to customize the advertised name if you want to access
+    # Kafka from outside the kube    
+    # Drop the SASL_PLAINTEXT endpoint if you want to allow only TLS (here and above)
+    # The Proxy uses the same TLS certificate configured for the Pulsar protocol
+    #    PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://pulsar-proxy:9092,SASL_SSL://pulsar-proxy:9093"
+    # if you use port forwarding to localhost then you have to set 'localhost'    
+    PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://pulsar-proxy:9092,SASL_SSL://pulsar-proxy:9093"
```
Comment (devinbost) [inline examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml:257]: @michaeljmarshall I want to find a way to abstract all this Kafka-specific configuration, as well as the Kafka configs in the proxy extensions section, into something that starts like:


```
starlight:
  kafka:
    enabled: true
    ssl: true
    . . .
```
However, it's not clear to me which of these parameters a customer would likely need to override. (How often do people really override these ports at this level instead of just using a load balancer to do that?)

**Comment by devinbost** [comment]
Superseded by https://github.com/datastax/pulsar-helm-chart/pull/278
